### PR TITLE
Add translation loader

### PR DESCRIPTION
**Pull request by Xiaochao8**
…juachon/cache-refactor

### CODE REVIEW

File: lib/sgtn-client/api/translation.rb [hunk 1]
```diff
@@ -67,97 +69,42 @@ def self.getTranslation(component, key, locale)
...
+        # TODO: Lock to initialize?
+        env = SgtnClient::Config.default_environment
+        mode = SgtnClient::Config.configurations[env]['bundle_mode']
+        SgtnClient.logger.debug "[Translation][init_translations]mode=#{mode}"
 
-      translation_bundle_thread = Thread.new { load(component, locale) }
-      old_source_bundle = load(component, LocaleUtil.get_source_locale)
-      source_bundle = get_cs(component, LocaleUtil.get_source_locale)
-      translation_bundle = translation_bundle_thread.value
+        if mode == 'offline'
```
Comment (jessiejuachon) [inline lib/sgtn-client/api/translation.rb:92]: How do you plan to support mixed mode (falling back to offline bundle if online bundle is not available)?

Comment (Xiaochao8) [reply lib/sgtn-client/api/translation.rb:92]: I have a rough idea now. Will create another PR for it.

File: spec/temp_spec.rb [hunk 2]
```diff
@@ -0,0 +1,18 @@
+# require 'sgtn-client/api/translation'
```
Comment (jessiejuachon) [inline spec/temp_spec.rb:1]: What is this file?

File: spec/single_operation/single_operation_spec.rb [hunk 3]
```diff
@@ -0,0 +1,4 @@
+# require 'sgtn-client/api/translation'
```
Comment (jessiejuachon) [inline spec/single_operation/single_operation_spec.rb:1]: Add license header

**Review by jessiejuachon — CHANGES_REQUESTED** [review]

File: lib/sgtn-client/api/translation.rb [hunk 4]
```diff
@@ -67,97 +69,42 @@ def self.getTranslation(component, key, locale)
...
-        cache_item = SgtnClient::CacheUtil.write_cache(cache_key, fetch(component, locale))
-        # delete thread from hash after finish
-        Thread.new { @refresh_cache_operator.remove_object(cache_key) }
-        cache_item
-      end
-    end
-    def self.refresh_cache(component, locale)
-      cache_key = SgtnClient::CacheUtil.get_cachekey(component, locale)
-      @refresh_cache_operator.operate(cache_key, component, locale)
+      def load_translations; end
```
Comment (jessiejuachon) [inline lib/sgtn-client/api/translation.rb:105]: what does this do?

**Review by jessiejuachon — CHANGES_REQUESTED** [review]

File: Gemfile [hunk 5]
```diff
@@ -16,4 +16,6 @@ group :test do
   gem 'rspec'
   gem 'webmock'
   gem 'pry-byebug', require: true
+  gem 'pry-inline', require: true
+  gem 'pry-doc'
```
[the remark this answers is not in the document]
Comment (Xiaochao8) [reply Gemfile:20]: These are good to debug. They are loaded in development env only.

File: lib/sgtn-client/api/translation.rb (hunk 4 above)
Comment (Xiaochao8) [reply lib/sgtn-client/api/translation.rb:105]: This is to load all translations(components and locales) to cache.

Comment (jessiejuachon) [reply lib/sgtn-client/api/translation.rb:105]: How does it load all translations if it is empty?

File: lib/sgtn-client/loader/cache.rb [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+  autoload :SingleOperation, 'sgtn-client/common/single_operation'
+end
+module SgtnClient::TranslationLoader::Cache
+  def load_bundle(component, locale)
+    cache_key = SgtnClient::CacheUtil.get_cachekey(component, locale)
+    SgtnClient.logger.debug "[#{self.to_s}][#{__FILE__}][#{__method__}] cache_key=#{cache_key}"
+    cache_item = SgtnClient::CacheUtil.get_cache(cache_key)
+    if cache_item.nil?
+      # refresh synchronously if not in cache
+      cache_item = (single_loader { |c, l| super(c, l) }).operate(cache_key, component, locale).value
```
Comment (jessiejuachon) [inline lib/sgtn-client/loader/cache.rb:15]:
```suggestion
      SgtnClient.logger.debug "[#{self.to_s}][#{__FILE__}][#{__method__}] Cache miss. cache_key=#{cache_key}"
      cache_item = (single_loader { |c, l| super(c, l) }).operate(cache_key, component, locale).value 
```

**Review by jessiejuachon — APPROVED** [review]

**Review by deminyan — APPROVED** [review]

**Pull request merged**